File: data/resources/user/user_model.py
from py2neo.ogm import Property, RelatedTo
from resources.model import BaseModel
from resources.movie.movie_model import MovieModel
from resources.review.review_model import ReviewModel


class UserModel(BaseModel):
    __primarykey__ = "id"
    __primarylabel__ = "User"

    id = Property()
    firstName = Property()
    lastName = Property()

    favourited = RelatedTo(MovieModel)
    wrote_review = RelatedTo(ReviewModel)

    def find_favourited(self):
        return self.find_related_to(self.favourited)

    def find_wrote_review(self):
        return self.find_related_to(self.wrote_review)

    def json(self):
        return {
            **dict(self.__node__),
            "favourited": self.find_favourited(),
            "wrote_review": self.find_wrote_review()
        }

File: server/src/resources/user/userModel.ts
import { Model } from "objection";
import { INeo4jMovie } from "../movie/MovieModel";
import { INeo4jReview } from "../review/ReviewModel";

export interface IAuthorizedUser {
  id: IUserModel["id"];
}
export interface IUserModel<Id = number> {
  id: Id;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  avatarURL?: string;
  avatarId?: string;
  isVerified: boolean;
}

export interface IAuthUser {
  id: string;
  firstName: string;
  lastName: string;
  avatarURL: string | undefined;
  isVerified: boolean;
  token: string;
}

export interface INeo4jUser {
  id: number;
  firstName: string;
  lastName: string;
  favourited: INeo4jMovie[];
  wrote_review: INeo4jReview[];
}

export interface IUserPayload {
  id: number;
  firstName: string;
  lastName: string;
}

export class UserModel extends Model implements IUserModel {
  id!: number;
  firstName!: string;
  lastName!: string;
  email!: string;
  password!: string;
  avatarURL?: string;
  avatarId?: string;
  isVerified!: boolean;

  $toData(): IUserModel {
    return {
      id: this.id,
      firstName: this.firstName,
      lastName: this.lastName,
      email: this.email,
      password: this.password,
      avatarURL: this.avatarURL,
      avatarId: this.avatarId,
      isVerified: this.isVerified,
    };
  }

  static get tableName() {
    return "Users";
  }

  static get jsonSchema() {
    return {
      type: "object",
      required: ["firstName", "lastName", "email", "password"],
      properties: {
        id: { type: "integer" },
        firstName: { type: "string", minLength: 1, maxLength: 255 },
        lastname: { type: "string", minLength: 1, maxLength: 255 },
        email: { type: "string", minLength: 1, maxLength: 255 },
        password: { type: "string", minLength: 1, maxLength: 255 },
        avatarURL: { type: "string", minLength: 1, maxLength: 510 },
        avatarId: { type: "string", minLength: 1, maxLength: 255 },
        isVerified: { type: "boolean" },
      },
    };
  }
}
